File: livebox/src/main/java/com/creations/livebox/serializers/LiveboxGsonSerializer.java
package com.creations.livebox.serializers;

import com.google.gson.Gson;
import com.google.gson.JsonIOException;
import com.google.gson.JsonSyntaxException;
import com.google.gson.stream.JsonReader;

import java.io.InputStreamReader;
import java.lang.reflect.Type;

import okio.BufferedSource;

import static com.creations.livebox.util.io.OkioUtils.bufferedSource;
import static com.creations.livebox.util.io.OkioUtils.readerInputStreamUtf8;

/**
 * @author Sérgio Serra on 25/08/2018.
 * Criations
 * sergioserra99@gmail.com
 */
public class LiveboxGsonSerializer implements Serializer {

    private Gson mGson;

    private LiveboxGsonSerializer(Gson gson) {
        mGson = gson;
    }

    public static LiveboxGsonSerializer create() {
        return new LiveboxGsonSerializer(new Gson());
    }

    public static LiveboxGsonSerializer create(Gson gson) {
        return new LiveboxGsonSerializer(gson);
    }

    @Override
    public <T> BufferedSource serialize(T input, Type type) {
        return bufferedSource(readerInputStreamUtf8(mGson.toJson(input, type)));
    }

    @Override
    public <T> T deserialize(BufferedSource source, Type type) {
        if (source == null) {
            return null;
        }

        try {
            return mGson.fromJson(new JsonReader(new InputStreamReader(source.inputStream())), type);
        } catch (JsonIOException e) {
            e.printStackTrace();
        } catch (JsonSyntaxException e) {
            e.printStackTrace();
        }

        return null;
    }
}

File: serializer-gson/src/main/java/com/creations/serializer_gson/LiveboxGsonSerializer.kt
package com.creations.serializer_gson

import com.creations.livebox_common.serializers.Serializer
import com.creations.livebox_common.util.bufferedSource
import com.creations.livebox_common.util.readerInputStreamUtf8
import com.google.gson.Gson
import com.google.gson.JsonIOException
import com.google.gson.JsonSyntaxException
import com.google.gson.stream.JsonReader
import okio.BufferedSource
import java.io.InputStreamReader
import java.lang.reflect.Type


/**
 * @author Sérgio Serra
 * sergioserra99@gmail.com
 */
class LiveboxGsonSerializer private constructor(private val mGson: Gson) : Serializer {

    override fun <T> serialize(input: T, type: Type): BufferedSource? {
        try {
            return bufferedSource(readerInputStreamUtf8(mGson.toJson(input, type)))
        } catch (e: JsonIOException) {
            e.printStackTrace()
        }
        return null
    }

    override fun <T> deserialize(source: BufferedSource, type: Type): T? {
        try {
            return mGson.fromJson<T>(JsonReader(InputStreamReader(source.inputStream())), type)
        } catch (e: JsonIOException) {
            e.printStackTrace()
        } catch (e: JsonSyntaxException) {
            e.printStackTrace()
        }
        return null
    }

    companion object {
        @JvmStatic
        fun create(gson: Gson = Gson()): Serializer = LiveboxGsonSerializer(gson)

        @JvmStatic
        fun create(): Serializer = create(Gson())
    }
}
